libs/data_providers_sdk/src/price_data/fetchers/fetch.rs

```
@@ -0,0 +1,35 @@
+use std::time::Instant;
+
+use anyhow::Result;
+use futures::stream::StreamExt;
+use futures::Stream;
+
+use crate::price_data::{traits::prices_fetcher::PairPriceData, types::ProviderPriceData};
+
+pub async fn fetch_all_prices<S>(mut futures_set: S) -> Result<Vec<ProviderPriceData>>
+where
+    S: Stream<Item = (&'static str, Result<PairPriceData>)> + Unpin,
+{
+    let before_fetch = Instant::now();
+    let mut all_fetched_prices: Vec<ProviderPriceData> = Vec::new();
+
+    // Process results as they complete
+    while let Some((exchange_id, result)) = futures_set.next().await {
+        match result {
+            Ok(prices) => {
+                let time_taken = before_fetch.elapsed();
+                println!("ℹ️  Successfully fetched prices from {exchange_id} in {time_taken:?}",);
+                let prices_per_exchange = ProviderPriceData {
+                    name: exchange_id.to_owned(),
+                    data: prices,
+                };
+                all_fetched_prices.push(prices_per_exchange);
+            }
+            Err(err) => println!("❌ Error fetching prices from {exchange_id}: {err:?}"),
```

We never return `Err` from this function.
Are we not supposed to error out on the first unsuccessful fetch?
Or maybe we should collect all the errors and return them instead of/alongside the successfully fetched prices?
If not, we could consider strengthening the return type to just `Vec<ProviderPriceData>`.

---

Also, general question, are plain `stdout` writes considered sufficient for error logging?

---

Let's just go with changing the return type. The main goal is to ensure that if any fetcher fails, it's logged in the oracle, but we don't need to handle the error beyond that. There might be different reasons of fails (e.g. hitting rate limit, slow connection, etc.), but we want to collect data from as many as possible providers. We don't want to return the errors.

As the oracle is being executed in `wasm` currently we don't support tracing and thats why we use `stdout`.